### PR TITLE
Cut 3473 GUI Progress Form

### PR DESCRIPTION
**Pull request by kmaranionjc**
## Issues
* [CUT-3473](https://jumpcloud.atlassian.net/browse/CUT-3473) -CUT-3473-GUI-Progress-Form

## What does this solve?
1. This change introduces new UI progress form that replaces the terminal verbose/progress. This solves the many information given to admin with the current verbosity thrown in the terminal which is confusing and unnecessary. The new form consists of progress bar, optional log text block, view log button, rerun JCAdmu button.
2. Update old Form logo to new one

### **OLD**
Form with terminal:
image
Migration:
image
image

### **NEW**
New ADMU Form Logo: 
image

Success:
image
Fail:
image

## How should this be tested?
Run ADMU Form like your normally would

**Success**
1.  Create .exe
3. When Migrate Profile is clicked, this should introduce the new progress form
4. Validate that the progress form is running normally as it would for a successful run
5. Validate all View Log, Rerun, and Exit buttons in the bottom are working properly

**Failure:**
1. Run ADMU
2. Lock a registry created by ADMU
3. Validate error is thrown and all buttons are working

**Rerun:**
1. After a migration or failure, rerun the app by pressing the Rerun button
2. Validate all functionality is working properly like when running the app from the beginning

**Terminal**
1. Run the terminal version of the tool and make sure that the terminal progress bar is working properly since I created a new function that consolidates both Write-Progress cmd and the new progress form update.

[CUT-3473]: https://jumpcloud.atlassian.net/browse/CUT-3473?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        # View Log Button
+        $SyncHash.ViewLog.Add_Click({
+            # Open log \Windows\Temp\jcAdmu.log
+            $scriptPath = "$(Get-WindowsDrive)\Windows\Temp\jcAdmu.log"
+            # Open the log
+           Invoke-Item -Path:($scriptPath)
+       })
+        # Start JCADMU Button
+        $syncHash.StartJCADMU.Add_Click({
+            Set-ExecutionPolicy Bypass -Scope Process -Force
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/ProgressForm.ps1:123]: Set-ExecutionPolicy? Why is this needed? I'd like to avoid changing this within a script

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    $newRunspace.SessionStateProxy.SetVariable("syncHash", $syncHash)
+
+    # Add the form code to the powershell instance
+    $psCommand = [PowerShell]::Create().AddScript({
+        # Load an xaml form
+
+        $syncHash.Window = [Windows.Markup.XamlReader]::parse( $SyncHash.XAML )
+        ([xml]$SyncHash.XAML).SelectNodes("//*[@Name]") | % { $SyncHash."$($_.Name)" = $SyncHash.Window.FindName($_.Name) }
+        # Image
+        $SyncHash.JCLogoImg.Source = $syncHash.base64JCLogo
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/ProgressForm.ps1:70]: I've seen at least a few iterations of running the script where this fails to load. We can probably cut down on the size of source image you've used here to improve performance.

Screenshot 2024-05-23 at 12 32 46 PM

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+        # TODO: Icon
+        #$SyncHash.Window.Icon = $syncHash.base64JCLogo
+        # Automatically scroll to the bottom of the log
+
+        # Scroll to end of Log
+        $syncHash.ScrollLog.ScrollToEnd()
+
+        $updateForm = {
+            $SyncHash.Window.Title = "JumpCloud ADMU 2.7.0"
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/ProgressForm.ps1:80]: Tests need to be updated to check if this title matches the current release version. We do this for `Form.ps1` and `Start-Migration.ps1` we should also be keeping ourselves honest here.

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            # IF close window button is clicked
+            if ($SyncHash.Closing) {
+                $SyncHash.Window.Close()
+                [System.Windows.Forms.Application]::Exit()
+                # Break out of the loop
+                Break
+                # Run Start-JCADMU
+            }
+            if ($SyncHash.logLevel -eq "Error") {
+                $syncHash.Status.Foreground = "Red"
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/ProgressForm.ps1:95]: I'm not sold on the red text here at the moment, we've talked about it but will move the conversation to the Jamboard.

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 5]
```diff
@@ -1,4 +1,5 @@
 #region Functions
+. "$PSScriptRoot\ProgressForm.ps1"
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Start-Migration.ps1:2]: Technically we do not need to load this when the program starts. I'd like you to come up with a more elegant way of selectively loading this series of functions. 

Completely related, this is only needed when we run the GUI version of the tool.

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 6]
```diff
@@ -2917,19 +3011,20 @@ Function Start-Migration {
         }
         if ([System.String]::IsNullOrEmpty($($admuTracker.Keys | Where-Object { $admuTracker[$_].fail -eq $true }))) {
             Write-ToLog -Message:('Script finished successfully; Log file location: ' + $jcAdmuLogFile) -Level Verbose
+            Write-ToProgress -ProgressBar $Progressbar -PercentComplete 100 -Status "Migration Completed Succesfully" -form $isForm
             Write-ToLog -Message:('Tool options chosen were : ' + "`nInstall JC Agent = " + $InstallJCAgent + "`nLeave Domain = " + $LeaveDomain + "`nForce Reboot = " + $ForceReboot + "`nUpdate Home Path = " + $UpdateHomePath + "`nAutobind JC User = " + $AutobindJCUser) -Level Verbose
-            if ($displayGuiPrompt) {
+            if (!$isForm) {
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Start-Migration.ps1:3016]: I believe there's an unintended regression here:

When you migrate with the production version `2.6.8` we do not display a message post-migration. This version of the module displays the message box even if the script is invoked from CLI.

Screenshot 2024-05-23 at 12 43 00 PM
Screenshot 2024-05-23 at 12 44 22 PM

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 7]
```diff
@@ -2917,19 +3011,20 @@ Function Start-Migration {
...
                 Show-Result -domainUser $SelectedUserName -localUser "$($localComputerName)\$($JumpCloudUserName)" -success $true -profilePath $newUserProfileImagePath -logPath $jcAdmuLogFile
             }
         } else {
             Write-ToLog -Message:("ADMU encoutered the following errors: $($admuTracker.Keys | Where-Object { $admuTracker[$_].fail -eq $true })") -Level Warn
             Write-ToLog -Message:("The following migration steps were reverted to their original state: $FixedErrors") -Level Warn
-            if ($displayGuiPrompt) {
+            Write-ToLog -Message:('Script finished with errors; Log file location: ' + $jcAdmuLogFile) -Level Error
+            Write-ToProgress -ProgressBar $Progressbar -PercentComplete 100 -Status $Script:ErrorMessage -form $isForm -logLevel "Error"
+            if (!$isForm) {
                 Show-Result -domainUser $SelectedUserName -localUser "$($localComputerName)\$($JumpCloudUserName)" -success $false -profilePath $newUserProfileImagePath -admuTrackerInput $admuTracker -FixedErrors $FixedErrors -logPath $jcAdmuLogFile
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Start-Migration.ps1:3025]: Show result can really be removed and condensed into the progress form.

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 8]
```diff
@@ -1843,13 +1901,73 @@ Function Start-Migration {
         [Parameter(ParameterSetName = "form")][Object]$inputObject)
 
     Begin {
+        $progressCount = 16
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Start-Migration.ps1:1904]: Where are the 16 items? I feel like this should be something dynamic, maybe we define an ordered list of steps in `Start-Migration` and pull from that list here. 

Editing this list seems really painful in the current implementation. I'd need to count the number of times we have a help message to display, then keep track of that number and go back to this function and manually update this number. Seems like we can just define an ordered hash table and pull the message contents from there?

**Review by jworkmanjc — CHANGES_REQUESTED** [review]
**Test items to address**
- Tests should be written to validate the "version" from `progressForm.ps1` matches the release version. These are currently missing.

**Functional items to address**

- The CLI version prompts with the `show-result` function. It shouldn't display anything, it should be silent.
- The EXE build is currently failing, this pull request should include changes to `Build-exe.ps1` to address the new code introduced. 
- Importing the file `ProgressForm.ps1` isn't going to be sufficient/ (may not be needed?) when the .EXE is generated. The process to generate the EXE is to dump all the PWSH code into one file `ADMU.ps1` and build an EXE from that single file. I'm okay to change this process but we need a way to bring your code from `ProgressForm.ps1` into the module if we've invoked the function from `form.ps1`.
- The logo is stretched, this looks off, I have a separate list of GUI changes to discuss. Sometimes the logo fails to load, we should modify the file to see if we can compress this and improve performance.

Overall I think this is a good step in the right direction, strong work here! There are a few things I wish were more elegant like the messaging, I'd like to see that better implemented. This is a large size ask of work I still think there's some clean up that we should do before merging this. 

We also need to make a decision on what to do with the `show-result` function. We need to either move part of it's useful messaging elements into the progress form or remove it. I believe it's no longer needed. 

I've got a running list of GUI changes we should be addressing.

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 (hunk 1 above)
Comment (kmaranionjc) [reply jumpcloud-ADMU/Powershell/ProgressForm.ps1:158]: Changed execution by file path to restart process

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+function DecodeBase64Image {
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/ProgressForm.ps1:2]: When we build the EXE, technically this function is duplicated in the ADMU.ps1 file.

I'm going to write a card to address this before we release 2.7.0.

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+        [Parameter(Mandatory = $true)]
+        $ProgressBar,
+        [string]$LogText
+    )
+    # Update the progress bar
+    $ProgressBar.LogText += $LogText
+}
+
+# $newProgress = New-ProgressForm
+# Update-ProgressForm -ProgressBar $newProgress -PercentComplete 50 -Status "Migration in progress" -logLevel "Info" -username "testuser" -profileSize "1GB" -localPath "C:\Users\testuser" -newLocalUsername "newtestuser"
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/ProgressForm.ps1:282]: remove comments

**Review by jworkmanjc — CHANGES_REQUESTED** [review]
few items to address, mostly code cleanup.

File: jumpcloud-ADMU/Powershell/ProgressForm.ps1 (hunk 9 above)
Comment (kmaranionjc) [reply jumpcloud-ADMU/Powershell/ProgressForm.ps1:2]: Agree, was thinking of commenting it out for local testing purposes in the future.

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 11]
```diff
@@ -1,5 +1,5 @@
 #region Functions
-
+#. $psScriptRoot\ProgressForm.ps1
```
[the remark this answers is not in the document]
Comment (kmaranionjc) [reply jumpcloud-ADMU/Powershell/Start-Migration.ps1:2]: Removed this in my last commit but vscode is acting up, will remove now

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 12]
```diff
@@ -1843,13 +1954,98 @@ Function Start-Migration {
...
+            $displayGuiPrompt = $true
+            Add-Type -AssemblyName System.Windows.Forms
+            # Use these
+            #     WindowStyle="SingleBorderWindow"
+            # ResizeMode="NoResize"
+            # Background="White" ScrollViewer.VerticalScrollBarVisibility="Visible" ScrollViewer.HorizontalScrollBarVisibility="Visible" Width="1000" Height="520">
+
+        }
+
+        $SelectedUserSid = Test-UsernameOrSID $SelectedUserName
```
[the remark this answers is not in the document]
Comment (kmaranionjc) [reply jumpcloud-ADMU/Powershell/Start-Migration.ps1:2039]: Added it to if($isForm){}else{$SelectedUserSid = Test-UsernameOrSID $SelectedUserName} instead of it being defined twice. Nice catch

**Review by jworkmanjc — APPROVED** [review]
Tested both GUI/ CLI versions, working as expected :) Nice work here. thank you for all of the changes!

**Review by gweinjc — APPROVED** [review]
Tested good on my machine and looks great! Awesome job!
